File: 0043_multiply_strings/multiply.py
def multiply(num1: str, num2: str) -> str:
    """
    Time: O(n * m)
    Space: O(n + m)
    """
    res = [0] * (len(num1) + len(num2))
    for i in reversed(range(len(num1))):
        for j in reversed(range(len(num2))):
            res[i + j + 1] += int(num1[i]) * int(num2[j])
            res[i + j] += res[i + j + 1] // 10
            res[i + j + 1] %= 10
    i = 0
    while i < len(res) and res[i] == 0:
        i += 1
    res = ''.join([str(ele) for ele in res[i:]])
    return res if res else '0'

File: 0043_multiply_strings/multiply.rb
# @param {String} num1
# @param {String} num2
# @return {String}
def multiply(num1, num2)
    # Time: O(n * m)
    # Space: O(n + m)
    res = Array.new(num1.length + num2.length + 1, 0)
    num1 = num1.reverse
    num2 = num2.reverse
    for i in (0...num1.length)
        for j in (0...num2.length)
            res[i + j] += num1[i].to_i * num2[j].to_i
            res[i + j + 1] += res[i + j]/10
            res[i + j] %= 10
        end
    end
    res = res.reverse
    start = 0
    start += 1 while res[start] == 0 && start < res.length - 1
    res[start..-1].join()
end